src/Functions/fromReadableSize.cpp

```diff
@@ -0,0 +1,265 @@
+                    {
+                        throw Exception(
+                            ErrorCodes::BAD_ARGUMENTS,
+                            "Invalid expression for function {}, find number after '.' failed, str: \"{}\".",
+                            getName(),
+                            String(str));
+                    }
+                }
+
+                /// convert float/integer string to float
+                Float64 base = 0;
+                std::string_view base_str = str.substr(token_front, token_tail - token_front);
+                auto value = boost::convert<Float64>(base_str, boost::cnv::strtol());
+                if (!value.has_value())
+                {
+                    throw Exception(
+                        ErrorCodes::BAD_ARGUMENTS,
+                        "Invalid expression for function {}, convert string to float64 failed: \"{}\".",
+                        getName(),
+                        String(base_str));
+                }
+                base = value.get();
+
+                scanSpaces(str, token_tail, last_pos);
+                token_front = token_tail;
+
+                /// scan a unit
+                if (!scanUnit(str, token_tail, last_pos))
+                {
+                    throw Exception(
+                        ErrorCodes::BAD_ARGUMENTS,
+                        "Invalid expression for function {}, find unit failed, str: \"{}\".",
+                        getName(),
+                        String(str));
+                }
+
+                std::string unit = std::string{str.substr(token_front, token_tail - token_front)};
+                boost::algorithm::to_lower(unit);
+                auto iter = size_unit_to_bytes.find(unit);
+                if (iter == size_unit_to_bytes.end()) /// not find unit
+                {
+                    throw Exception(
+                        ErrorCodes::BAD_ARGUMENTS, "Invalid expression for function {}, parse unit failed: \"{}\".", getName(), unit);
+                }
+                Float64 raw_num_bytes = base * iter->second;
+                if (raw_num_bytes > MAX_UINT64)
+                {
+                    throw Exception(
+                        ErrorCodes::BAD_ARGUMENTS,
+                        "Invalid expression for function {}, result is too big for output data type (UInt64): \"{}\".",
+                        getName(),
+                        raw_num_bytes
+                    );
+                }
+                // As the input might be an arbitrary decimal number we might end up with a non-integer amount of bytes when parsing binary (eg MiB) units.
+                // This doesn't make sense so we round up to indicate the byte size that can fit the passed size.
+                result = static_cast<UInt64>(std::ceil(raw_num_bytes));
+
+                res_data.emplace_back(result);
+            }
+
+            return col_to;
+        }
+
+        /// scan an unsigned integer number
```

Comment:
The following functions could be in the `private` section of the class.

Reply:
After reworking the parser implementation to use the functions in `IO/ReadHelpers` none of those helpers are needed anymore so I've got rid of them